From 2f67b4e53495a8e338aef6a0efb8dd06c654e599 Mon Sep 17 00:00:00 2001
From: xuxiang <xuxiang@eswincomputing.com>
Date: Fri, 17 May 2024 14:26:48 +0800
Subject: [PATCH 038/109] WIN2030-15279:fix:spi_slv access 4G + space

Changelogs:
1. spi slv write/read addr more than 4G address space

Change-Id: I6c2bc08868b7ef763138e2d9c8ea6c39b140a0cc
---
 Core/Inc/hf_spi_slv.h |  4 +-
 Core/Src/hf_spi_slv.c | 95 ++++++++++++++++++++-----------------------
 2 files changed, 45 insertions(+), 54 deletions(-)

diff --git a/Core/Inc/hf_spi_slv.h b/Core/Inc/hf_spi_slv.h
index 86c91d0..95f021d 100644
--- a/Core/Inc/hf_spi_slv.h
+++ b/Core/Inc/hf_spi_slv.h
@@ -15,8 +15,8 @@
 		HAL_GPIO_WritePin(SPI2_NSS_GPIO_Port, SPI2_NSS_Pin, GPIO_PIN_SET); \
 	} while (0);
 
-int es_spi_write(unsigned char *buf, unsigned long addr, int len);
-int es_spi_read(unsigned char *dst, unsigned long src, int len);
+int es_spi_write(uint8_t *buf, uint64_t addr, int len);
+int es_spi_read(uint8_t *dst, uint64_t src, int len);
 
 #ifdef __cplusplus
 }
diff --git a/Core/Src/hf_spi_slv.c b/Core/Src/hf_spi_slv.c
index a364a7e..12e1d24 100644
--- a/Core/Src/hf_spi_slv.c
+++ b/Core/Src/hf_spi_slv.c
@@ -3,7 +3,7 @@
 #include "hf_spi_slv.h"
 #include "stm32f4xx_hal.h"
 
-unsigned int get_regval(unsigned char reg);
+uint32_t get_regval(uint8_t reg);
 
 HAL_StatusTypeDef spi_transmit(uint8_t *pData, uint16_t Size)
 {
@@ -35,7 +35,7 @@ HAL_StatusTypeDef spi_recive(uint8_t *pData, uint16_t Size)
 	SPI2_FLASH_CS_HIGH();
 	return ret;
 }
-void spi_addr_cfg(unsigned char *cmd, unsigned long addr)
+void spi_addr_cfg(uint8_t *cmd, uint64_t addr)
 {
 	cmd[5] = (addr & 0xff);
 	cmd[4] = ((addr >> 8) & 0xff);
@@ -45,10 +45,9 @@ void spi_addr_cfg(unsigned char *cmd, unsigned long addr)
 	cmd[0] = ((addr >> 40) & 0xff);
 }
 
-void spi_data_cfg(unsigned char *cmd, unsigned char *data, int len)
+void spi_data_cfg(uint8_t *cmd, uint8_t *data, int len)
 {
 	int i;
-	unsigned char xdata;
 	switch (len) {
 	case 1:
 		cmd[0] = data[0];
@@ -70,51 +69,51 @@ void spi_data_cfg(unsigned char *cmd, unsigned char *data, int len)
 	}
 }
 
-int eswin_ww(unsigned long addr, unsigned int data)
+int eswin_ww(uint64_t addr, uint32_t data)
 {
-	unsigned char sndBuf[32];
+	uint8_t sndBuf[32];
 	HAL_StatusTypeDef ret;
 	// AXI word write: cmd(1B)+addr(6B)+data(4B)
 	sndBuf[0] = 0x1;
 	spi_addr_cfg(&sndBuf[1], addr);
-	spi_data_cfg(&sndBuf[7], (unsigned char *)&data, 4);
+	spi_data_cfg(&sndBuf[7], (uint8_t *)&data, 4);
 	ret = spi_transmit(sndBuf, 11);
 	return ret;
 }
 
-int eswin_wb(unsigned long addr, unsigned char data)
+int eswin_wb(uint64_t addr, uint8_t data)
 {
 	int ret;
-	unsigned char sndBuf[32];
+	uint8_t sndBuf[32];
 
 	// AXI word write: cmd(1B)+addr(6B)+data(1B)
 	sndBuf[0] = 0xb;
 	spi_addr_cfg(&sndBuf[1], addr);
-	spi_data_cfg(&sndBuf[7], (unsigned char *)&data, 1);
+	spi_data_cfg(&sndBuf[7], (uint8_t *)&data, 1);
 
 	ret = spi_transmit(sndBuf, 8);
 	return ret;
 }
 
-int eswin_wh(unsigned long addr, unsigned short data)
+int eswin_wh(uint64_t addr, unsigned short data)
 {
 	int ret;
-	unsigned char sndBuf[32];
+	uint8_t sndBuf[32];
 
 	// AXI word write: cmd(1B)+addr(6B)+data(2B)
 	sndBuf[0] = 0x9;
 	spi_addr_cfg(&sndBuf[1], addr);
-	spi_data_cfg(&sndBuf[7], (unsigned char *)&data, 2);
+	spi_data_cfg(&sndBuf[7], (uint8_t *)&data, 2);
 
 	ret = spi_transmit(sndBuf, 9);
 	return ret;
 }
 
-int eswin_burst_write(unsigned long addr, unsigned char *data)
+int eswin_burst_write(uint64_t addr, uint8_t *data)
 {
 	// AXI burst write: cmd(1B)+addr(6B)+data(32B)
 	int ret;
-	unsigned char sndBuf[39];
+	uint8_t sndBuf[39];
 
 	sndBuf[0] = 0x3;
 	spi_addr_cfg(&sndBuf[1], addr);
@@ -125,10 +124,10 @@ int eswin_burst_write(unsigned long addr, unsigned char *data)
 
 
 
-void format_big_2_little(unsigned char *buf, int len)
+void format_big_2_little(uint8_t *buf, int len)
 {
-	unsigned char data;
-	int left = len, i = 0;
+	uint8_t data;
+	int i = 0;
 
 	if (len == 1)
 		return;
@@ -155,10 +154,10 @@ void format_big_2_little(unsigned char *buf, int len)
 	}
 }
 
-int spi_axi_read(unsigned char *rcvBuf, int len)
+int spi_axi_read(uint8_t *rcvBuf, int len)
 {
-	int ret;
-	unsigned char sndBuf;
+	int ret = HAL_OK;
+	uint8_t sndBuf;
 
 	if (len == 1) {
 		sndBuf = 0x1a;
@@ -167,16 +166,14 @@ int spi_axi_read(unsigned char *rcvBuf, int len)
 	} else if ((len == 4) || (len == 32)) {
 		sndBuf = 0x16;
 	}
-	ret = spi_transmit_recive(sndBuf, 1, rcvBuf, len);
+	ret = spi_transmit_recive(&sndBuf, 1, rcvBuf, len);
 	return ret;
 }
 
-int eswin_rx(unsigned char *rcvBuf, unsigned long addr, int len)
+int eswin_rx(uint8_t *rcvBuf, uint64_t addr, int len)
 {
-	int ret, xlen, offset = 0;
-	unsigned char sndBuf[32];
-	unsigned int rvdata = 0;
-	printf("%s %x %x\n",__func__, addr, len);
+	int ret = HAL_OK, xlen, offset = 0;
+	uint8_t sndBuf[32];
 	while (len) {
 		if (len >= 32) {
 			sndBuf[0] = 0x2;
@@ -197,9 +194,6 @@ int eswin_rx(unsigned char *rcvBuf, unsigned long addr, int len)
 		}
 
 		spi_addr_cfg(&sndBuf[1], addr);
-		for(int i = 0; i<7; i++) {
-			printf("sndBuf[%d] %x\n", i, sndBuf[i]);
-		}
 		if (xlen == 1) {
 			sndBuf[7] = 0x1a;
 		} else if (xlen == 2) {
@@ -223,23 +217,22 @@ fail:
 	return ret;
 }
 
-int es_spi_writeb(unsigned long addr, unsigned char val)
+int es_spi_writeb(uint64_t addr, uint8_t val)
 {
 	return eswin_wb(addr, val);
 }
-int es_spi_writew(unsigned long addr, unsigned short val)
+int es_spi_writew(uint64_t addr, unsigned short val)
 {
 	return eswin_wh(addr, val);
 }
-int es_spi_writel(unsigned long addr, unsigned int val)
+int es_spi_writel(uint64_t addr, uint32_t val)
 {
 	return eswin_ww(addr, val);
 }
 
-int es_spi_write(unsigned char *buf, unsigned long addr, int len)
+int es_spi_write(uint8_t *buf, uint64_t addr, int len)
 {
-	int xlen, offset = 0, ret;
-	printf("addr %x len %x\n",addr, len);
+	int xlen, offset = 0, ret = HAL_OK;
 	while (len) {
 		if (len >= 32) {
 			len -= 32;
@@ -250,7 +243,7 @@ int es_spi_write(unsigned char *buf, unsigned long addr, int len)
 		} else if (len >= 4) {
 			xlen = 4;
 			len -= 4;
-			ret = es_spi_writel(addr, *(unsigned int *)(buf + offset));
+			ret = es_spi_writel(addr, *(uint32_t *)(buf + offset));
 			if (ret)
 				goto fail;
 		} else if (len >= 2) {
@@ -274,44 +267,42 @@ fail:
 	return ret;
 }
 
-unsigned char es_spi_readb(unsigned long addr)
+uint8_t es_spi_readb(uint64_t addr)
 {
-	unsigned char buf[1];
+	uint8_t buf[1];
 	eswin_rx(buf, addr, 1);
 
 	return *buf;
 }
 
-unsigned short es_spi_readw(unsigned long addr)
+unsigned short es_spi_readw(uint64_t addr)
 {
-	unsigned char buf[2];
+	uint8_t buf[2];
 	eswin_rx(buf, addr, 2);
 
 	return *(unsigned short *)buf;
 }
 
-unsigned int es_spi_readl(unsigned long addr)
+uint32_t es_spi_readl(uint64_t addr)
 {
-	unsigned char buf[4];
+	uint8_t buf[4];
 	eswin_rx(buf, addr, 4);
 
-	return *(unsigned int *)buf;
+	return *(uint32_t *)buf;
 }
 
-int es_spi_read(unsigned char *dst, unsigned long src, int len)
+int es_spi_read(uint8_t *dst, uint64_t src, int len)
 {
 	return eswin_rx(dst, src, len);
 }
 
-unsigned int get_regval(unsigned char reg)
+uint32_t get_regval(uint8_t reg)
 {
-	unsigned char sndBuf[2];
-	unsigned char rcvBuf[4];
-	int ret;
+	uint8_t sndBuf[2];
+	uint8_t rcvBuf[4] = {0};
 	sndBuf[0] = 0x4;
 	sndBuf[1] = reg;
-	ret = spi_transmit_recive(sndBuf, 2, rcvBuf, 4);
+	spi_transmit_recive(sndBuf, 2, rcvBuf, 4);
 	format_big_2_little(rcvBuf,4);
-	// printf("reg %x 0x%08x\n",reg>>4,*(unsigned int *)rcvBuf);
-	return *(unsigned int *)rcvBuf;
+	return *(uint32_t *)rcvBuf;
 }
\ No newline at end of file
-- 
2.25.1

